File: packages/core/src/research_engine/domain/passages.py
"""Passage and search domain types."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from research_engine.domain.common import FusionMode


class Passage(BaseModel):
    """A chunk of a document — the unit of retrieval."""

    id: UUID
    document_id: UUID
    position: int
    char_start: int | None = None
    char_end: int | None = None
    locator: dict[str, Any] = Field(default_factory=dict)
    text: str
    token_count: int | None = None
    chunker: str
    chunker_version: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    #: The structural node containing this passage, when the document has a
    #: tree. None for corpora ingested before nodes existed.
    node_id: UUID | None = None
    content_hash: bytes
    created_at: datetime


class PassageDraft(BaseModel):
    """Data needed to create a passage record.

    ``char_start`` / ``char_end`` are the passage's span in the document's
    canonical text, and are required: they are the address every other feature
    hangs off — pin-cites, quote verification, annotations, re-chunking. The
    contract every chunker must satisfy is::

        draft.text == canonical_text[draft.char_start:draft.char_end]

    ``locator`` stays for type-specific extras (page, verse, timecode) that are
    meaningful to a reader but not usable as an address.
    """

    position: int
    char_start: int
    char_end: int
    locator: dict[str, Any] = Field(default_factory=dict)
    text: str
    token_count: int | None = None
    chunker: str
    chunker_version: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    #: Resolved after the document tree is written, since node ids do not
    #: exist while a chunker is running.
    node_id: UUID | None = None

    @model_validator(mode="after")
    def _span_is_well_formed(self) -> PassageDraft:
        if self.char_start < 0:
            raise ValueError(f"char_start must be non-negative, got {self.char_start}")
        if self.char_end < self.char_start:
            raise ValueError(
                f"char_end ({self.char_end}) precedes char_start ({self.char_start})"
            )
        if self.char_end - self.char_start != len(self.text):
            raise ValueError(
                f"span width {self.char_end - self.char_start} does not match "
                f"text length {len(self.text)} — the span and the text disagree"
            )
        return self


class PassageWindow(BaseModel):
    """The expanded read of a hit — more than matched, bounded by structure.

    Not what was ranked. The chunk is, and it is still on the hit as ``text``.
    This is what a person or an agent should *read*: a chunk boundary is where
    the ingester happened to cut, which in a lexicon lands mid-definition and in
    a monograph mid-argument.

    Keeps the same invariant ``PassageDraft`` keeps —
    ``text == canonical_text[char_start:char_end]`` — so a window can be quoted
    and re-located exactly like a passage can.
    """

    text: str
    char_start: int
    char_end: int
    #: How the boundaries were chosen. ``node`` means the window *is* one
    #: structural node and `read_node` would add nothing; anything else means it
    #: is a slice and reading the node would give more.
    source: Literal["node", "node_window", "document_window", "passage"]
    #: The node that bounded the window — not always the passage's own node. A
    #: node narrower than the chunk gets widened inside its parent instead.
    node_id: UUID | None = None
    #: Ancestor titles, root first. The citation for this window.
    breadcrumb: list[str] = Field(default_factory=list)
    #: Measured on the returned text, not on the estimate that sized it.
    approx_tokens: int


class PassageHit(BaseModel):
    """A passage returned by search with scores."""

    passage_id: UUID
    document_id: UUID
    score: float
    score_breakdown: ScoreBreakdown | None = None
    #: The chunk that actually matched — what was embedded, ranked and reranked.
    #: Quote this. Read ``window``.
    text: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    locator: dict[str, Any] = Field(default_factory=dict)
    #: The chunk's span in the document's canonical text. Carried so a caller can
    #: widen a hit without re-reading the row; ``None`` on passages written
    #: before offsets were required.
    char_start: int | None = None
    char_end: int | None = None
    node_id: UUID | None = None
    #: ``None`` when the document has no canonical text to widen into, or the
    #: passage has no offsets to widen from.
    window: PassageWindow | None = None


class ScoreBreakdown(BaseModel):
    """Breakdown of how the score was computed."""

    vector: float | None = None
    keyword: float | None = None
    rerank: float | None = None
    rrf: float | None = None


class SearchQuery(BaseModel):
    """Search query with filters and hybrid options."""

    text: str
    filters: SearchFilters | None = None
    k: int = 20
    k_vec: int = 100
    k_kw: int = 100
    fusion_mode: FusionMode = FusionMode.rrf
    alpha: float = 0.5
    rerank: bool = True
    rerank_n: int = 30


class SearchFilters(BaseModel):
    """Filters for narrowing search results."""

    document_types: list[str] | None = None
    date_range_start: str | None = None
    date_range_end: str | None = None
    author_entity_id: UUID | None = None
    recipient_entity_id: UUID | None = None
    mentions_entity_ids: list[UUID] | None = None
    metadata: dict[str, Any] | None = None
    language: str | None = None
    extensions: dict[str, Any] | None = None
    extension_logic: Literal["and", "or"] = "and"


class SearchResult(BaseModel):
    """Result of a search operation."""

    hits: list[PassageHit]
    total_candidates: int
    applied_filters: dict[str, Any] = Field(default_factory=dict)
    #: Stages that were skipped because a backend was unreachable, e.g.
    #: ``["rerank_unavailable"]``. Empty means the search ran in full.
    #:
    #: This exists so that degrading is never silent. Reranking is 99.4% of
    #: query latency, so when the GPU host is down the only sensible thing is to
    #: return fused results without it — but results ordered by RRF alone are
    #: measurably different from reranked ones, and a researcher comparing
    #: today's hits against last week's deserves to know which they are looking
    #: at rather than inferring it from a mood.
    degraded: list[str] = Field(default_factory=list)

File: crates/marginalia-types/src/passages.rs
//! Passage and search types mirroring `domain/passages.py`.
//!
//! `PassageDraft` is defined in [`crate::sdk`] — core re-exports the SDK type
//! exactly as `domain/passages.py` re-exports `research_engine_sdk`'s — and is
//! re-exported here for the same reason.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

pub use crate::sdk::PassageDraft;

/// A chunk of a document — the unit of retrieval.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Passage {
    pub id: Uuid,
    pub document_id: Uuid,
    pub position: i64,
    #[serde(default)]
    pub char_start: Option<i64>,
    #[serde(default)]
    pub char_end: Option<i64>,
    #[serde(default)]
    pub locator: Map<String, Value>,
    pub text: String,
    #[serde(default)]
    pub token_count: Option<i64>,
    pub chunker: String,
    pub chunker_version: String,
    #[serde(default)]
    pub metadata: Map<String, Value>,
    /// The structural node containing this passage. None for corpora ingested
    /// before nodes existed.
    #[serde(default)]
    pub node_id: Option<Uuid>,
    #[serde(with = "crate::wire::bytes_string")]
    pub content_hash: Vec<u8>,
    pub created_at: DateTime<Utc>,
}

/// The expanded read of a hit — more than matched, bounded by structure.
///
/// Keeps the `PassageDraft` invariant `text == canonical[char_start:char_end]`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PassageWindow {
    pub text: String,
    pub char_start: i64,
    pub char_end: i64,
    /// `node` means the window *is* one structural node; anything else means
    /// it is a slice and reading the node would give more.
    pub source: WindowSource,
    /// The node that bounded the window — not always the passage's own node.
    #[serde(default)]
    pub node_id: Option<Uuid>,
    /// Ancestor titles, root first. The citation for this window.
    #[serde(default)]
    pub breadcrumb: Vec<String>,
    /// Measured on the returned text, not on the estimate that sized it.
    pub approx_tokens: i64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WindowSource {
    Node,
    NodeWindow,
    DocumentWindow,
    Passage,
}

/// What a hit is cited from — the citation draft for a search result.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct HitSource {
    #[serde(default)]
    pub document_title: Option<String>,
    /// The bibliographic join, once a pack writes it at ingest. None until then.
    #[serde(default)]
    pub edition_key: Option<String>,
    /// `documents.metadata.edition`, when a pack wrote one.
    #[serde(default)]
    pub edition: Option<String>,
    /// `document_texts.parser_version`. None when there is no canonical text.
    #[serde(default)]
    pub parser_version: Option<String>,
    #[serde(default)]
    pub has_canonical_text: bool,
    /// False on rows from older chunkers lacking offsets: not citable.
    #[serde(default)]
    pub has_offsets: bool,
}

/// A passage returned by search with scores.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PassageHit {
    pub passage_id: Uuid,
    pub document_id: Uuid,
    pub score: f64,
    #[serde(default)]
    pub score_breakdown: Option<ScoreBreakdown>,
    #[serde(default)]
    pub source: Option<HitSource>,
    /// The chunk that actually matched — what was embedded, ranked, reranked.
    pub text: String,
    #[serde(default)]
    pub metadata: Map<String, Value>,
    #[serde(default)]
    pub locator: Map<String, Value>,
    /// Carried so a caller can widen a hit without re-reading the row.
    #[serde(default)]
    pub char_start: Option<i64>,
    #[serde(default)]
    pub char_end: Option<i64>,
    #[serde(default)]
    pub node_id: Option<Uuid>,
    /// None when the document has no canonical text, or the passage no offsets.
    #[serde(default)]
    pub window: Option<PassageWindow>,
}

/// Breakdown of how the score was computed.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct ScoreBreakdown {
    #[serde(default)]
    pub vector: Option<f64>,
    #[serde(default)]
    pub keyword: Option<f64>,
    #[serde(default)]
    pub rerank: Option<f64>,
    #[serde(default)]
    pub rrf: Option<f64>,
}

/// Search query with filters and hybrid options.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SearchQuery {
    pub text: String,
    #[serde(default)]
    pub filters: Option<SearchFilters>,
    #[serde(default = "default_k")]
    pub k: i64,
    #[serde(default = "default_k_vec")]
    pub k_vec: i64,
    #[serde(default = "default_k_kw")]
    pub k_kw: i64,
    #[serde(default)]
    pub fusion_mode: crate::common::FusionMode,
    #[serde(default = "default_alpha")]
    pub alpha: f64,
    #[serde(default = "default_true")]
    pub rerank: bool,
    #[serde(default = "default_rerank_n")]
    pub rerank_n: i64,
}

fn default_k() -> i64 {
    20
}
fn default_k_vec() -> i64 {
    100
}
fn default_k_kw() -> i64 {
    100
}
fn default_alpha() -> f64 {
    0.5
}
fn default_true() -> bool {
    true
}

fn default_rerank_n() -> i64 {
    30
}

/// Filters for narrowing search results.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct SearchFilters {
    #[serde(default)]
    pub document_types: Option<Vec<String>>,
    /// Raw strings in core (parsed downstream), unlike the SDK's datetimes.
    #[serde(default)]
    pub date_range_start: Option<String>,
    #[serde(default)]
    pub date_range_end: Option<String>,
    #[serde(default)]
    pub author_entity_id: Option<Uuid>,
    #[serde(default)]
    pub recipient_entity_id: Option<Uuid>,
    #[serde(default)]
    pub mentions_entity_ids: Option<Vec<Uuid>>,
    #[serde(default)]
    pub metadata: Option<Map<String, Value>>,
    #[serde(default)]
    pub language: Option<String>,
    #[serde(default)]
    pub extensions: Option<Map<String, Value>>,
    #[serde(default = "default_extension_logic")]
    pub extension_logic: ExtensionLogic,
}

fn default_extension_logic() -> ExtensionLogic {
    ExtensionLogic::And
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ExtensionLogic {
    #[default]
    And,
    Or,
}

/// Result of a search operation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SearchResult {
    pub hits: Vec<PassageHit>,
    pub total_candidates: i64,
    #[serde(default)]
    pub applied_filters: Map<String, Value>,
    #[serde(default)]
    pub degraded: Vec<String>,
}
